qutebrowser/commands/runners.py

```diff
@@ -59,58 +66,67 @@ def _current_url(tabbed_browser):
-                                window=win_id)
-    url = lambda: _current_url(tabbed_browser)
-
-    variables = {
-        'url': lambda: url().toString(
+def _init_variable_replacements() -> typing.Dict[str, ReplacementFunction]:
+    """Return a dict from variable replacements to fns processing them."""
+    url = _current_url
+    replacements = {
+        'url': lambda tb: url(tb).toString(
             QUrl.FullyEncoded | QUrl.RemovePassword),
-        'url:pretty': lambda: url().toString(
+        'url:pretty': lambda tb: url(tb).toString(
             QUrl.DecodeReserved | QUrl.RemovePassword),
-        'url:domain': lambda: "{}://{}{}".format(
-            url().scheme(),
-            url().host(),
-            ":" + str(url().port()) if url().port() != -1 else ""),
-        'url:auth': lambda: "{}:{}@".format(
-            url().userName(),
-            url().password()) if url().userName() else "",
-        'url:scheme': lambda: url().scheme(),
-        'url:username': lambda: url().userName(),
-        'url:password': lambda: url().password(),
-        'url:host': lambda: url().host(),
-        'url:port': lambda: str(url().port()) if url().port() != -1 else "",
-        'url:path': lambda: url().path(),
-        'url:query': lambda: url().query(),
-        'title': lambda: tabbed_browser.widget.page_title(
-            tabbed_browser.widget.currentIndex()),
-        'clipboard': utils.get_clipboard,
-        'primary': lambda: utils.get_clipboard(selection=True),
-    }
-
-    for key in list(variables):
+        'url:domain': lambda tb: "{}://{}{}".format(
+            url(tb).scheme(), url(tb).host(),
+            ":" + str(url(tb).port()) if url(tb).port() != -1 else ""),
+        'url:auth': lambda tb: "{}:{}@".format(
+            url(tb).userName(),
+            url(tb).password()) if url(tb).userName() else "",
+        'url:scheme': lambda tb: url(tb).scheme(),
+        'url:username': lambda tb: url(tb).userName(),
+        'url:password': lambda tb: url(tb).password(),
+        'url:host': lambda tb: url(tb).host(),
+        'url:port': lambda tb: str(
+            url(tb).port()) if url(tb).port() != -1 else "",
+        'url:path': lambda tb: url(tb).path(),
+        'url:query': lambda tb: url(tb).query(),
+        'title': lambda tb: tb.widget.page_title(
+            tb.widget.currentIndex()),
+        'clipboard': lambda _: utils.get_clipboard(),
+        'primary': lambda _: utils.get_clipboard(selection=True),
+    }  # type:typing.Dict[str, ReplacementFunction]
+    for key in list(replacements):
         modified_key = '{' + key + '}'
-        variables[modified_key] = lambda x=modified_key: x
+        # x = modified_key is to avoid binding x as a closure
+        replacements[modified_key] = (
+            lambda _, x=modified_key: x)  # type: ignore
+    return replacements
+
 
+VARIABLE_REPLACEMENTS = types.MappingProxyType(_init_variable_replacements())
```

Comment:
Why the `MappingProxyType`?

Reply:
Likely to ensure that it is not accidentally changed in some other places.

Reply:
`Mapping` is just a more loose version of `Dict` (like `Sequence` or `Iterable` compared to `List`), I don't think there's currently a way to make mypy check immutability.

As for using `types.MappingProxyType`: I don't think this makes sense here. The name clearly conveys that it's constant, so if it'd be modified somewhere else, that'd stand out anyways. That kind of stuff IMHO makes sense where accidental mutations can happen easily (like with objects coming from the config), but I don't think that's the case here.

Reply:
No, `MutableMapping` is a more loose version of `Dict`. `Mapping` is immutable.

Reply:
There's also `from typing_extensions import Final` (see https://github.com/python/mypy/pull/5522 for usage), but that is another kind of immutability - that the variable cannot be reassigned. It's not the case here.